### DISCUSSION TITLE
Do retrying requests require lots of memory?

### DISCUSSION THREAD

**Discussion by LilTwo** [vectordotdev/vector#17522] [Q&A]
Our team has encountered OOM crashed of vector running on k8s several times,
after reading the document about [buffering model](https://vector.dev/docs/about/under-the-hood/architecture/buffering-model/),
we got confused because we have all the buffer settings to be default
```
buffer.type = "memory"
buffer.max_events = 500
buffer.when_full = "block"
```
(we didn't actually write these in the config)

With a Kafka source which has average event size to be ~300 bytes, we expected the max memory usage of vector should be around
300 bytes * 500 # = 150 kb
and when buffer is full, source should stop consuming message from Kafka
but the memory usage reached over 1Gb, and messages are still consumed from Kafka until the pod was OOM killed instead
So we did a experiment, set `buffer.max_events` to 1 and have a http sink that does not exist, and vector still consumed more than 10000 events from Kafka
 
after digging into the source code, I found that sender has its own memory buffer:
https://github.com/vectordotdev/vector/blob/98c54ad3a371ac710151367a953252f9eb293548/src/topology/builder.rs#L61

it can buffer thousands of events before sending to sink memory buffer, and this buffer size seems to be unconfigurable

We also got many logs like this
`WARN sink{component_kind="sink" component_id=sink component_type=aws_kinesis_streams component_name=sink}:request{request_id=1977859}: vector::sinks::util::retries: Retrying after error. error=TransientError: connection closed before message completed internal_log_rate_limit=true`

Seems that requests were retried over and over again while events were still consumed from the Kafka source
Would this be the reason that we got such high memory usage?

As my current understanding there are at least 3 buffers for a event to go through before sending out
source buffer => sink buffer => sink batching buffer
where do the events waited for retrying stand?

And it would be really helpful if there are a page describes all the settings that could affect the memory usage of a vector process, thanks🙏

**Comment by LinTechSo** [comment]
Hi, 
any update on this ?
